read_buffer/src/row_group.rs

```
@@ -743,129 +811,85 @@ impl RowGroup {
+            for (group_col_i, col) in group_cols_out.iter_mut().enumerate() {
+                let decoded_value =
+                    input_group_columns[group_col_i].decode_id(group_key_buf[group_col_i] as u32);
 
-            // TODO(edd): given these RLE columns should have low cardinality
-            // there should be a reasonably low group key cardinality. It could
-            // be safe to use `small_vec` here without blowing the stack up.
-            let mut material_key = Vec::with_capacity(group_key.len());
-            for (col_idx, &encoded_id) in group_key.iter().enumerate() {
-                material_key.push(group_columns[col_idx].decode_id(encoded_id as u32));
+                col.push(match decoded_value {
+                    Value::Null => None,
+                    Value::String(s) => Some(s),
+                    _ => panic!("currently unsupported group column"),
+                });
             }
-            dst.group_keys.push(GroupKey(material_key));
 
-            let mut aggregates = Vec::with_capacity(aggregate_columns_typ.len());
-            for (agg_col, typ) in &aggregate_columns_typ {
-                aggregates.push(match typ {
+            // Calculate an aggregate from each input aggregate column and
+            // set it at the relevant offset in the output column.
+            for (agg_col_i, (agg_col, typ)) in input_aggregate_columns.iter().enumerate() {
+                match typ {
                     AggregateType::Count => {
-                        AggregateResult::Count(agg_col.count(&aggregate_row_ids.to_vec()) as u64)
+                        let agg = agg_col.count(&group_key_row_ids.to_vec()) as u64;
+                        agg_cols_out[agg_col_i].push(Value::Scalar(Scalar::U64(agg)))
                     }
-                    AggregateType::First => todo!(),
-                    AggregateType::Last => todo!(),
+                    AggregateType::First => {}
```

these will be filled out in a future PR. They can't be executed yet as there is error checking at the read buffer API to reject first/last

---

is there a reason to leave them as `{}` and not `unimplemented!()` or some other `panic!` inducing macro to fail fast (rather than failing in a more silent and possibly harder to debug way)?

---

I think it was just a side-effect of my editor filling out arms. I'm going to be working on these implementations this week anyhow.